File: mvc/src/main/java/com/example/icecream/di/assessment_record/ModelModule.java
package com.example.icecream.di.assessment_record;

import android.os.Bundle;

import com.example.firebasedb.RemoteStorage;
import com.example.icecream.di.PerActivity;
import com.example.icecream.model.AssessmentRecordsManager;
import com.example.icecream.model.AssessmentRecordsManagerImpl;
import com.example.icecream.model.ObservableAssessmentRecord;
import com.example.icecream.network.NetworkStateMonitor;

import dagger.Module;
import dagger.Provides;

@Module
public class ModelModule {

    private Bundle savedState;

    public ModelModule(Bundle savedState) {
        this.savedState = savedState;
    }

    @Provides
    @PerActivity
    ObservableAssessmentRecord provideObservableAssessmentRecord() {
        return new ObservableAssessmentRecord();
    }

    @Provides
    @PerActivity
    AssessmentRecordsManager provideAssessmentRecordsManager(NetworkStateMonitor networkStateMonitor,
                                                             RemoteStorage remoteStorage) {
        return new AssessmentRecordsManagerImpl(networkStateMonitor, remoteStorage, savedState);
    }
}

File: mvvm/src/main/java/com/example/icecream/di/assessment_record/ModelModule.kt
package com.example.icecream.di.assessment_record

import android.os.Bundle
import com.example.firebasedb.RxFirebaseRemoteStorage
import com.example.firebasedb.RxRemoteStorage
import com.example.icecream.di.PerActivity
import com.example.icecream.model.Model
import com.example.icecream.model.ModelImpl
import dagger.Module
import dagger.Provides

@Module
class ModelModule(private val savedState: Bundle?) {

    @Provides
    @PerActivity
    internal fun provideModel(remoteStorage: RxRemoteStorage): Model {
        return ModelImpl(remoteStorage, savedState)
    }

    @Provides
    @PerActivity
    internal fun provideRxRemoteStorage(): RxRemoteStorage {
        return RxFirebaseRemoteStorage()
    }
}
